ui-v2/app/controllers/settings.js

```diff
@@ -5,8 +5,19 @@ import { inject as service } from '@ember/service';
 export default Controller.extend({
   repo: service('settings'),
   dom: service('dom'),
+  wait: service('timeout'),
+  confirming: false,
+  applyTransition: function() {
+    const wait = get(this, 'wait').execute;
+    set(this, 'confirming', true);
+    // TODO: Make 0 default in wait
+    wait(0).then(() => {
```

Comment:
The ember idiom for this is [`run.next`](https://api.emberjs.com/ember/2.15/namespaces/Ember.run/methods/next?anchor=next), which also integrates with Backburner.

```js
import { next } from '@ember/runloop';
// ...
set(this, 'confirming', true);
next(() => {
  set(this, 'confirming', false);
});
```

Reply:
Hey!

So I moved to use `next`/`backburner` instead of `setTimeout` 7772dd4 . Then thought I'd like to keep this injectable _and_ follow a Promise API. So I made like a cross between the two fce53c4, which gives us:

```javascript
set(this, 'confirming', true);
tick().then(() => {
  set(this, 'confirming', false);
});
```

or

```javascript
set(this, 'confirming', true);
await tick()
set(this, 'confirming', false);
```

I also considered just:

```javascript
// following the set api, but also potentially use .then/await if you want to wait for it.
tick(this, 'confirming', true);
```

...which would set it to `true` and then back to whatever it was before the tick. I'd probably like to add that in future.

What do you think of each? (`await tick()` and `tick(this, 'confirming', true)`)

I keep reading that Octane will 'eliminate the runloop', but I'm not 100% sure if they just mean 'the runloop' or `@ember/runloop` and the things inside it like `next` etc? Any idea? This is half the reason I'd like to inject this.

Ideally I'd move this throughout the app where I'm using `wait(0)` in a couple of other places (I actually have a todo to pretty much do this). But thought I'd do that as a separate task/PR, post getting your feedback on ^.

Reply:
> What do you think of each? (await tick() and tick(this, 'confirming', true))

I like `await tick()`. The `tick()` one-liner is also nice, but not immediately obvious it's asynchronous.

> I keep reading that Octane will 'eliminate the runloop', but I'm not 100% sure if they just mean 'the runloop' or @ember/runloop and the things inside it like next etc? Any idea? This is half the reason I'd like to inject this.

Hmm. I haven't heard about this but I'd welcome it 😛 Ember has a history of shipping codemods along with these types of changes, so I personally don't think there's harm in using `next` even if it's on the chopping block.

Reply:
Re: runloop things, I think that's what this post is talking about: https://discuss.emberjs.com/t/readers-questions-why-does-ember-still-use-rsvp/14736 - in 3.4 and above, RSVP is essentially equivalent to the native Promise implementation (previously it wasn't because of incorrect browser implementations). The runloop isn't really going away, but the autorun assertion is removed so you don't have to fiddle with it in tests (and eventually RSVP could probably be removed, but there's not currently plans to do that).

Reply:
> I like await tick(). The tick() one-liner is also nice, but not immediately obvious it's asynchronous.

Yeah I know what you mean, but I do like that 'usually' you don't want to wait for the nextTick here:

```javascript
tick(this, 'confirming', true)
// do something else
// I probably don't care that tick ^ hasn't finished yet
```

```javascript
await tick(this, 'confirming', true)
// this time for some reason I do care, so I .then/await it
```
Kind of 'fire and forget', but if you want to `.then/await` it you can.

Anyway, I'll leave that for a potential future PR, we'd really like to start animating some more things in the UI and this will be a nice little tool to help us.

---

Hey Matthew :wave: ! Yeah RSVP is also going 🎉 , but specifically the 'runloop' also 'I think' 🤷‍♂ 

Oh yeah good link! This is interesting:

> Does this mean the run loop is going away? Not really – it’s still a thing that you want to exist behind the scenes, reliably making sure there’s always a render scheduled to happen at the right time to keep your work optimally batched. But the runloop as something that app developers will need to learn about just to get their app working & tested can definitely go away, and that will make me very happy.

Maybe this statement still stands? Maybe its changed slightly, it's dated May 2018

Here's the Octane list as it stands right now (crossed out things are things they've since decided not to do yet as far as I understand)

'Part 1'

> - Broccoli 2.0 in ember-cli
> - ~~Module Unification~~ - De-scoped from this edition as per https://blog.emberjs.com/2019/03/11/update-on-module-unification-and-octane.html
> - Glimmer Components as the default component API
> - Native JavaScript Classes as the default object model
> - ~~Native JavaScript modules~~ - De-scoped from this edition exposing modules in the build pipeline publishing Ember as npm packages importing npm packages

'Part 2'

> - No jQuery by default
> - ~~Svelte Builds (moved to next edition)~~
> - Native JavaScript Classes
> - Glimmer components
> - ~~Incremental rendering and rehydration (moved to next edition)~~
> - ~~Treeshaking~~ (moved to next edition)
> - Eliminating the runloop
> - ~~Stabilizing Ember Data (moved to next edition)~~

Just realized they've repeated some of those, so this is the full list:

> - Broccoli 2.0 in ember-cli
> - ~~Module Unification~~ - De-scoped from this edition as per https://blog.emberjs.com/2019/03/11/update-on-module-unification-and-octane.html
> - Glimmer Components as the default component API
> - Native JavaScript Classes as the default object model
> - ~~Native JavaScript modules~~ - De-scoped from this edition exposing modules in the build pipeline publishing Ember as npm packages importing npm packages
> - No jQuery by default
> - ~~Svelte Builds~~ (moved to next edition)
> - ~~Incremental rendering and rehydration~~ (moved to next edition)
> - ~~Treeshaking~~ (moved to next edition)
> - Eliminating the runloop
> - ~~Stabilizing Ember Data (moved to next edition)~~

I'm really excited about Octane, and I really like that the emphasis seems to be on 'just javascript'

I would hazard a guess that 'eliminating the runloop' could mean removing Backburner and using raf or something simpler based on raf. I'm not sure if it means removing things like `next` though 🤷‍♂ . If you come across anything related to this I'd be super interested to read. 'Eliminating' sounds very much like 'we are completely getting rid', but who knows?

The other interesting thing to note here with the current work we are doing is:

> - Glimmer Components as the default component API

@DingoEatingFuzz and I chatted briefly about this the other day, but worth pulling out here also.

Thanks for the info and feedback folks! I'm gonna merge this, will possibly refine this in another PR with the single liner `tick` at a later date.

John